lib/rspec/core/reporter.rb

```diff
@@ -2,6 +2,15 @@ module RSpec::Core
   # A reporter will send notifications to listeners, usually formatters for the
   # spec suite run.
   class Reporter
+    # @private
+    NOTIFICATIONS =
+      [
+        :close, :deprecation, :deprecation_summary, :dump_failures, :dump_pending,
+        :dump_profile, :dump_summary, :example_failed, :example_group_finished,
+        :example_group_started, :example_passed, :example_pending, :example_started,
+        :message, :seed, :start, :start_dump, :stop
+      ]
```

Comment:
> I didn't want to use Set because it requires us to require it, but I've realised we already do for the deprecation formatter... I have mixed feelings about that, I know the performance is better but I'd really like to avoid requiring parts of the standard library, it's something that can catch us out.

`Set` is a simple wrapper around `Hash` which is what really gives constant time inclusion checks, so we can just use a `Hash` instead (e.g. with a value of `true` for each key or whatever...the values don't actually matter).  We should probably look into doing that for our other uses of `Set` although that's out of scope for this PR IMO.  We may also want to create our own wrapper class that wraps a `Hash` and provides the tiny bit of the `Set` interface we actually use.

> Personally I'd like to avoid the overhead of having to check each time in notify, but I'd be willing to work up a test to check that they're all included somehow?

A test would be cool but I couldn't think of a way to test that because such a test would need a list of which events to test, but that list itself is what we are actually trying to test.  I came up with this, though...it's basically the idea I mentioned before, but implemented in a prepended module that is defined in `spec/support/formatter_support` so as to only be active in test environments and not add any runtime overhead (besides to rspec-core's spec suite).

``` ruby
if RSpec::Support::RubyFeatures.module_prepends_supported?
  module RSpec::Core
    class Reporter
      module EnforceRSpecNotificationsListComplete
        def notify(event, *args)
          return super if caller_locations(1, 1).first.label =~ /publish/
          return super if RSPEC_NOTIFICATIONS.include?(event)

          raise "#{event.inspect} must be added to `RSPEC_NOTIFICATIONS`"
        end
      end

      prepend EnforceRSpecNotificationsListComplete
    end
  end
end
```

> I've changed it to RSPEC_NOTIFICATIONS if you think that helps?

:+1: